src/make_trade.ts

```diff
@@ -15,28 +24,70 @@ async function makeTrade(
     throw "Network doesn't expose a chainId";
   }
 
-  const all_tokens = await fetchTokenList(tokenListUrl, network.config.chainId);
-  const tokens_with_balance = await tokensWithBalance(
-    all_tokens,
+  const allTokens = await fetchTokenList(tokenListUrl, network.config.chainId);
+  const tokensWithBalance = await filterTokensWithBalance(
+    allTokens,
     trader,
     ethers
   );
-  if (tokens_with_balance.length === 0) {
+  if (tokensWithBalance.length === 0) {
     throw "Account doesn't have any balance in any of the provided token";
   }
 
-  const { token: sell_token, balance: sell_balance } = selectRandom(
-    tokens_with_balance
+  const { token: sellToken, balance: sellBalance } = selectRandom(
+    tokensWithBalance
+  );
+  const buyToken = selectRandom(
+    allTokens.filter((token) => sellToken !== token)
+  );
+
+  const api = new Api(network.name);
+  const fee = await api.getFee(
+    sellToken.address,
+    buyToken.address,
+    sellBalance,
+    OrderKind.SELL
   );
-  const buy_token = selectRandom(
-    all_tokens.filter((token) => sell_token !== token)
+
+  if (sellBalance.lte(fee)) {
+    throw "Account doesn't have enough balance to pay fee";
+  }
+
+  const sellAmountAfterFee = sellBalance.sub(fee);
+  const buyAmount = await api.estimateTradeAmount(
+    sellToken.address,
+    buyToken.address,
+    sellAmountAfterFee,
+    OrderKind.SELL
   );
 
   console.log(
-    `Selling ${sell_balance.toString()} of ${sell_token.name} for ${
-      buy_token.name
-    }`
+    `Selling ${sellAmountAfterFee.toString()} of ${
+      sellToken.name
+    } for ${buyAmount} of ${buyToken.name} with a ${fee.toString()} fee`
+  );
+
+  const order = createOrder(
+    sellToken,
+    buyToken,
+    sellAmountAfterFee,
+    buyAmount,
+    fee
+  );
+  const signature = await signOrder(
+    domain(
+      network.config.chainId,
+      // TODO: How do I get this address from the repo?
+      "0x4E608b7Da83f8E9213F554BDAA77C72e125529d0"
+    ),
+    order,
+    trader,
+    // TODO: TypedData does not seem to work with the ethers provider
+    SigningScheme.MESSAGE
```

Comment:
In general, it does. I imagine the `trader` signer being used doesn't provide a `_TypedDataSigner` implementation.

Reply:
Hmm, what exactly is not working? I tried this locally and I was able to sign.

Reply:
Is this issue the same as in this TODO? https://github.com/gnosis/gp-swap-ui/blob/639d025a5ae1be189209e3639c248b4be2e46f79/src/custom/utils/signatures.ts#L25-L43

That typescript can't export enums?

Reply:
I think the issue is the `hardhat` `ethers.Signer` implementation doesn't support `eth_signTypedData_v4` calls.